core/device.cpp

```diff
@@ -197,3 +199,102 @@ extern "C" void free_device_table(struct device_table *devices)
 {
 	delete devices;
 }
+
+// managing fingerprint data
+bool fingerprint_record::operator<(const fingerprint_record &a) const
+{
+	if (model == a.model)
+		return serial < a.serial;
+	return model < a.model;
+}
+
+// annoyingly, the Cressi Edy doesn't support a serial number (it's always 0), but still uses fingerprints
+// so we can't bail on the serial number being 0
+extern "C" unsigned int get_fingerprint_data(const struct fingerprint_table *table, uint32_t model, uint32_t serial, unsigned char **fp_out)
+{
+	if (model == 0 || fp_out == nullptr)
+		return 0;
+	struct fingerprint_record fpr = { model, serial };
+	auto it = std::lower_bound(table->fingerprints.begin(), table->fingerprints.end(), fpr);
+	if (it != table->fingerprints.end()) {
+		if (has_dive(it->fdeviceid, it->fdiveid)) {
+			*fp_out = it->raw_data;
+			return it->fsize;
+		}
+	}
+	return 0;
+}
+
+extern "C" void create_fingerprint_node(struct fingerprint_table *table, uint32_t model, uint32_t serial,
+				       const unsigned char *raw_data_in, unsigned int fsize, uint32_t fdeviceid, uint32_t fdiveid)
+{
+	// since raw data can contain \0 we copy this manually, not as string
+	unsigned char *raw_data = (unsigned char *)malloc(fsize);
+	if (!raw_data)
+		return;
+	memcpy(raw_data, raw_data_in, fsize);
+
+	struct fingerprint_record fpr = { model, serial, raw_data, fsize, fdeviceid, fdiveid };
+	auto it = std::lower_bound(table->fingerprints.begin(), table->fingerprints.end(), fpr);
+	if (it != table->fingerprints.end()) {
+		// update what we have
+		it->fdeviceid = fdeviceid;
+		it->fdiveid = fdiveid;
+		it->raw_data = raw_data;
+		it->fsize = fsize;
+	} else {
+		// insert a new one
+		table->fingerprints.insert(it, fpr);
+	}
+}
+
+extern "C" void create_fingerprint_node_from_hex(struct fingerprint_table *table, uint32_t model, uint32_t serial,
+						const char *hex_data, uint32_t fdeviceid, uint32_t fdiveid)
+{
+	QByteArray raw = QByteArray::fromHex(hex_data);
+	create_fingerprint_node(table, model, serial,
+				(const unsigned char *)raw.constData(), raw.size(), fdeviceid, fdiveid);
+}
+
+extern "C" int nr_fingerprints(struct fingerprint_table *table)
+{
+	return table->fingerprints.size();
+}
+
+extern "C" uint32_t fp_get_model(struct fingerprint_table *table, int i)
+{
+	if (!table || i < 0 || i >= table->fingerprints.size())
+		return 0;
+	return table->fingerprints.at(i).model;
+}
+
+extern "C" uint32_t fp_get_serial(struct fingerprint_table *table, int i)
+{
+	if (!table || i < 0 || i >= table->fingerprints.size())
+		return 0;
+	return table->fingerprints.at(i).serial;
+}
+
+extern "C" uint32_t fp_get_deviceid(struct fingerprint_table *table, int i)
+{
+	if (!table || i < 0 || i >= table->fingerprints.size())
+		return 0;
+	return table->fingerprints.at(i).fdeviceid;
+}
+
+extern "C" uint32_t fp_get_diveid(struct fingerprint_table *table, int i)
+{
+	if (!table || i < 0 || i >= table->fingerprints.size())
+		return 0;
+	return table->fingerprints.at(i).fdiveid;
```

Comment:
For your interest: in std::vector, the difference between `at(i)` and `[i]` is that the former will throw an exception for out-of-bound accesses, whereas the latte will gladly do read/write beyond the buffer. This may or may not be what you want.

Reply:
so you are saying I could simply use `table->fingerprints[i].fdiveid` here since I checked the bounds right above?

Reply:
I think both version are perfectly fine - this was just a FYI. [] would probably be a bit more idiomatic, since it doesn't usually result in an application termination on out-of-bound access.

Reply:
well, I switched to that in the last force push